### PR TITLE
Set Java 9 checkstyle to depend on checkstyle conf

### LINKED ISSUE #28376: Fix task ordering for checkstyle Java 9
The checkstyle configuration here is broken:
 - we are not configuring checkstyle for Java 9
 - the source set for Java 9 is not configured until after we have configured the checkstyle task

This can lead to bad ordering:
```
:server:checkstyleJava9 SKIPPED
:server:copyCheckstyleConf SKIPPED
:server:checkstyleMain SKIPPED
```

so the `:server:checkstyleJava9` can fail with

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':server:checkstyleJava9'.
> Unable to create Root Module: configLocation {/home/jason/src/elastic/elasticsearch/server/build/checkstyle/checkstyle.xml}, classpath {/home/jason/src/elastic/elasticsearch/server/build/classes/java/java9:/home/jason/src/elastic/elasticsearch/server/build/resources/java9:/home/jason/src/elastic/elasticsearch/server/build/classes/java/main:/home/jason/src/elastic/elasticsearch/server/build/resources/main}.
```

### PR DESCRIPTION
**Pull request by jasontedor**
We need to configure the Java 9 checkstyle task to depend on the checkstyle configuration task or the task could run before the checkstyle conf has been copied leading to runtime failures. We have to do this after projects have been evaluated because the configuration of these tasks can occur before the Java 9 source set has been added to a project.

Closes #28376

### CODE REVIEW

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy [hunk 1]
```diff
@@ -130,27 +130,29 @@ class PrecommitTasks {
...
-        for (String taskName : ['checkstyleMain', 'checkstyleTest']) {
-            Task task = project.tasks.findByName(taskName)
-            if (task != null) {
-                project.tasks['check'].dependsOn.remove(task)
-                checkstyleTask.dependsOn(task)
-                task.dependsOn(copyCheckstyleConf)
-                task.inputs.file(checkstyleSuppressions)
-                task.reports {
-                    html.enabled false
+        project.afterEvaluate {
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy]: I don't think we need an afterEvaluate. Instead, I think we can use:

```project.tasks.withType(Checkstyle) { Checkstyle checkstyleTask ->
  dependsOn(copyCheckstyleConf)
}```

Comment (jasontedor) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy]: I might be misunderstanding but I do not think that works; the problem here, and the reason for the `afterEvaluate` is that the checkstyleJava9 task is not defined when we create the precommit tasks. The source set for Java 9 is added later. This is why I do this in an `afterEvaluate`, to ensure that the Java 9 source set is added before the checkstyle tasks are configured.

Comment (rjernst) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy]: Many of the methods that gradle has on collections, like `withType`, operate not only on the current objects in that collection, but also on future objects added. So the closure applied to withType here will also be applied to the java9 checkstyle when it is later added.

**Comment by jasontedor** [comment]
@rjernst I have updated this pull request; would you please take another look?

**Review by hub-cap — APPROVED** [review]

**Pull request merged**